finality-aleph/src/new_network/split.rs

```diff
@@ -0,0 +1,230 @@
+}
+
+#[derive(Clone)]
+struct HidariSender<D1: Data, D2: Data, S: SenderComponent<Split<D1, D2>>> {
+    sender: S,
+    d1: PhantomData<D1>,
+    d2: PhantomData<D2>,
+}
+
+impl<D1: Data, D2: Data, S: SenderComponent<Split<D1, D2>>> SenderComponent<D1>
+    for HidariSender<D1, D2, S>
+{
+    fn send(&self, data: D1, recipient: Recipient) -> Result<(), SendError> {
+        self.sender.send(Split::Hidari(data), recipient)
+    }
+}
+
+#[derive(Clone)]
+struct MigiSender<D1: Data, D2: Data, S: SenderComponent<Split<D1, D2>>> {
+    sender: S,
+    d1: PhantomData<D1>,
+    d2: PhantomData<D2>,
+}
+
+impl<D1: Data, D2: Data, S: SenderComponent<Split<D1, D2>>> SenderComponent<D2>
+    for MigiSender<D1, D2, S>
+{
+    fn send(&self, data: D2, recipient: Recipient) -> Result<(), SendError> {
+        self.sender.send(Split::Migi(data), recipient)
+    }
+}
+
+struct HidariReceiver<D1: Data, D2: Data, R: ReceiverComponent<Split<D1, D2>>> {
+    receiver: Arc<Mutex<R>>,
+    translated_receiver: mpsc::UnboundedReceiver<D1>,
+    hidari_sender: mpsc::UnboundedSender<D1>,
+    migi_sender: mpsc::UnboundedSender<D2>,
+}
+
+struct MigiReceiver<D1: Data, D2: Data, R: ReceiverComponent<Split<D1, D2>>> {
+    receiver: Arc<Mutex<R>>,
+    translated_receiver: mpsc::UnboundedReceiver<D2>,
+    hidari_sender: mpsc::UnboundedSender<D1>,
+    migi_sender: mpsc::UnboundedSender<D2>,
+}
+
+async fn forward_or_wait<D1: Data, D2: Data, R: ReceiverComponent<Split<D1, D2>>>(
+    receiver: &Arc<Mutex<R>>,
+    hidari_sender: &mpsc::UnboundedSender<D1>,
+    migi_sender: &mpsc::UnboundedSender<D2>,
+) {
+    match receiver.try_lock().ok() {
+        Some(mut receiver) => match receiver.next().await {
+            Some(Split::Hidari(data)) => {
+                if hidari_sender.unbounded_send(data).is_err() {
+                    warn!(target: "aleph-network", "Failed send despite controlling receiver, this shouldn't've happened.");
+                }
+            }
+            Some(Split::Migi(data)) => {
+                if migi_sender.unbounded_send(data).is_err() {
+                    warn!(target: "aleph-network", "Failed send despite controlling receiver, this shouldn't've happened.");
+                }
+            }
+            None => {
+                hidari_sender.close_channel();
+                migi_sender.close_channel();
+            }
+        },
+        None => {
+            interval(Duration::from_millis(MAXIMUM_RETRY_MS))
+                .tick()
+                .await;
+        }
+    }
+}
+
+const MAXIMUM_RETRY_MS: u64 = 500;
+
+#[async_trait::async_trait]
+impl<D1: Data, D2: Data, R: ReceiverComponent<Split<D1, D2>>> ReceiverComponent<D1>
+    for HidariReceiver<D1, D2, R>
+{
+    async fn next(&mut self) -> Option<D1> {
+        loop {
+            tokio::select! {
+                data = self.translated_receiver.next() => return data,
+                _ = forward_or_wait(&self.receiver, &self.hidari_sender, &self.migi_sender) => (),
```

Comment:
It will always eventually go to the first case – either `forward_or_wait` will forward something there, or it will get closed, closing the channel and making `translated_receiver` return `None`.

Reply:
ok, makes sense now